spec-abstract-noclosure/ffg.tla

```diff
@@ -0,0 +1,265 @@
+        \* equivocation
+        \/ /\ vote1.validator = vote2.validator
+           /\ vote1 /= vote2
+           /\ vote1.ffg_vote.target[2] = vote2.ffg_vote.target[2]
+        \* surround voting
+        \/ /\ vote1.validator = vote2.validator
+           /\ \/ vote1.ffg_vote.source[2] < vote2.ffg_vote.source[2]
+              \/ /\ vote1.ffg_vote.source[2] = vote2.ffg_vote.source[2]
+                 /\ vote1.ffg_vote.source[1].slot < vote2.ffg_vote.source[1].slot
+           /\ vote2.ffg_vote.target[2] < vote1.ffg_vote.target[2]
+    } IN { v.validator: v \in slashable_votes }
+
+\* Append a block to chain 1.
+\* @type: Int => Bool;
+ProposeBlockOnChain1(slot) ==
+    LET new_block == Block(slot, NextBody(chain1_tip, "main")) IN
+    /\ IsValidBlock(new_block)
+    /\ slot > chain1_tip.slot
+    /\ all_blocks' = all_blocks \union { new_block }
+    /\ chain1' = chain1 \union { new_block }
+    /\ chain1_tip' = new_block
+    /\ IF IsForked
+       THEN UNCHANGED <<chain2, chain2_tip>>
+       ELSE
+         /\ chain2' = chain2 \union { new_block }
+         /\ chain2_tip' = new_block
+    /\ UNCHANGED <<ffg_votes, votes, justified_checkpoints, chain2_fork_block_number>>
+
+\* Append a block to chain 2.
+\* @type: Int => Bool;
+ProposeBlockOnChain2(slot) ==
+    LET new_block == Block(slot, NextBody(chain2_tip, "fork")) IN
+    /\ IsValidBlock(new_block)
+    /\ slot > chain2_tip.slot
+    /\ chain2_fork_block_number' =
+        IF IsForked THEN chain2_fork_block_number ELSE new_block.body
+    /\ all_blocks' = all_blocks \union { new_block }
+    /\ chain2' = chain2 \union { new_block }
+    /\ chain2_tip' = new_block
+    /\ UNCHANGED <<chain1, chain1_tip, ffg_votes, votes, justified_checkpoints>>
+
+JustifiedCheckpoints(viewVotes) ==
+    \* @type: Set($checkpoint) => Set($checkpoint);
+    LET AccJustified(justifiedSoFar, justifiedCheckpointSlot) ==
+        LET candidateCheckpoints == { Checkpoint(block, justifiedCheckpointSlot): block \in all_blocks } IN
+        LET newJustifiedCheckpoints == { c \in candidateCheckpoints: IsJustified(c, viewVotes, justifiedSoFar) } IN
+        justifiedSoFar \union newJustifiedCheckpoints
+    IN ApaFoldSeqLeft(AccJustified, { GenesisCheckpoint }, MkSeq(MAX_BLOCK_SLOT+2, (* @type: Int => Int; *) LAMBDA i: i))
+
+\* @type: ($checkpoint, $checkpoint, Set(Str)) => Bool;
+CastVotes(source, target, validators) ==
+    LET ffgVote == [ source |-> source, target |-> target ] IN
+    /\ IsValidFFGVote(ffgVote)
+    /\ validators /= {}
+    /\ ffg_votes' = ffg_votes \union { ffgVote }
+    /\ votes' = votes \union { Vote(v, ffgVote): v \in validators }
+    /\ LET allCheckpoints == {Checkpoint(block, i): block \in all_blocks, i \in CheckpointSlots}
+       IN \E allJustifiedCheckpoints \in SUBSET allCheckpoints:
+        /\ justified_checkpoints' = allJustifiedCheckpoints
+        /\ \A c \in allJustifiedCheckpoints: IsJustified(c, votes', allJustifiedCheckpoints)
```

Comment:
I am not sure about the role of `allJustifiedCheckpoints` in the predicate `IsJustified(c, votes', allJustifiedCheckpoints)` as `allJustifiedCheckpoints` and `fixpoint` seem not a good match. Should we pass the unprimed variable `justified_checkpoints`, instead of `allJustifiedCheckpoints`?